### DISCUSSION TITLE
How to check starlette websocket status??

### DISCUSSION THREAD

**Discussion by wjs2063** [fastapi/fastapi#11248] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.websocket("/{client_id}/ws")
@inject
async def websocket_test(*, client_id: str, websocket: WebSocket,
                         connection_manager: ConnectionManager = Depends(Provide[ClientContainer.connection_manager]),
                         external_manager: ConnectionManager = Depends(
                             Provide[ExternalContainer.external_connection_manager])
                         ):
    try:
        client_websocket = await connection_manager.connect(client_id=client_id, websocket=websocket)
        external_websocket = await external_manager.connect(client_id=client_id, websocket=websocket)
    except Exception as e:
        logger.info(e)

    try:
        text = await client_websocket.receive_text()
        await external_websocket.send(text)
    except Exception as e:
        print(e)
        logger.info(e)
        return
    try:
        while True:
            external_resp = await external_websocket.recv()
            external_resp = json.loads(external_resp)
            await client_websocket.send_json(external_resp)
    except websockets.exceptions.ConnectionClosed as e:
        err_resp = {
            "status_code": f"ConnectionClosed, {e.code}"
        }
        print(e)
    except websockets.exceptions.ConnectionClosedOK as e:
        err_resp = {
            "status_code": f"ConnectionClosedOK {e.code}"
        }
        print(e)
    except websockets.exceptions.ConnectionClosedError as e:
        err_resp = {
            "status_code": f"ConnectionClosedError {e.code}"
        }
        print(e)
    except websockets.exceptions.WebSocketException as e:
        err_resp = {
            "status_code": f"WebSocketException {e.code}"
        }
        print(e)
    except Exception as e:
        print(f"Internal Server Error : {e}")
        err_resp = {
            "status_code": f"{e}"
        }
    finally:
        await external_manager.disconnect(client_id=client_id, websocket=external_websocket)
        await connection_manager.disconnect(client_id=client_id, websocket=client_websocket)
```

lots of exception hadling here

### Description

```
[RuntimeError: Unexpected ASGI message ‘websocket.close’, after sending ‘websocket.close’
```

 how can I check the current state of the websocket object?? For example, whether it is connected or closed, etc.

I can't find status_function in here https://github.com/encode/starlette/blob/master/starlette/websockets.py

The reason for looking for this is to handle various errors.

### Operating System

macOS

### FastAPI Version

0.109.0

### Pydantic Version

2.6.0

### Python Version

Python 3.11.4

**Comment by Kludex** [comment]
What errors?

**Reply by wjs2063** [reply]
websocket is a bidirectional protocol , so server doesn't know when closed connection 
so i have to manage error , 

The basic logic of WebSocket would be accept(), send(), and receive(). Isn't it difficult to find out when and what error occurred? Is there any other way?

To be more specific, my situation is like this.
You need to manage two WebSocket objects within one endpoint. One is a web socket connection object from the client, and the other is a web socket connection with external_server.

I update my code

**Reply by Kludex** [reply]
I need an MRE.
